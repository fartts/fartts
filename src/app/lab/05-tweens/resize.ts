import { on } from '../../../lib/dom';

const { isInteger } = Number;

function next(a: number, b: number): number {
  while (!isInteger(a / b)) {
    a += 1;
  }

  return a;
}

let shouldResize = true;

export default function resize(
  c: HTMLCanvasElement,
  m: HTMLMainElement,
): boolean {
  if (!shouldResize) {
    return false;
  }

  shouldResize = false;

  let { devicePixelRatio: dpr } = window;
  let { clientWidth: width, clientHeight: height } = m;

  if (dpr === undefined) {
    dpr = 1;
  }

  const scale = 1;
  width = next(width, scale);
  height = next(height, scale);

  const w = (width * dpr) / scale;
  const h = (height * dpr) / scale;

  c.style.width = `${width}px`;
  c.style.height = `${height}px`;

  c.width = w;
  c.height = h;

  return true;
}

on('resize', () => {
  shouldResize = true;
});
