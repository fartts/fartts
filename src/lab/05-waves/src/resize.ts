import { on } from '../../../lib/core/dom';
import { max } from '../../../lib/core/math';

function next(multipleOf = 1, startingAt = 0) {
  return startingAt % multipleOf !== 0
    ? startingAt + (multipleOf - (startingAt % multipleOf))
    : startingAt;
}

let didResizeWindow = true;
on('resize', () => {
  didResizeWindow = true;
});

export function shouldResize() {
  return didResizeWindow;
}

export function resize(
  canvas: HTMLCanvasElement,
  container: HTMLElement,
  scale = 1,
): void {
  if (!didResizeWindow) {
    return;
  }

  didResizeWindow = false;

  const { devicePixelRatio: dpr } = window;
  const { clientHeight: h, clientWidth: w } = container;

  const width = next(scale, w);
  const height = next(scale, h);

  canvas.width = (width * dpr) / scale;
  canvas.height = (height * dpr) / scale;

  canvas.style.transform = `scale(${max(w / canvas.width, h / canvas.height)})`;
}
