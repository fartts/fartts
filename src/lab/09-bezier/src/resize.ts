import { on } from '../../../lib/core/dom';
import { max } from '../../../lib/core/math';

function nextMultiple(multipleOf = 1, startingFrom = 0) {
  return startingFrom % multipleOf !== 0
    ? startingFrom + (multipleOf - (startingFrom % multipleOf))
    : startingFrom;
}

let didResizeWindow = true;
on('resize', () => {
  didResizeWindow = true;
});

export function shouldResize() {
  return didResizeWindow;
}

export function resize(
  canvas: HTMLCanvasElement,
  container: HTMLElement,
  scale = 1,
): void {
  if (!didResizeWindow) {
    return;
  }

  didResizeWindow = false;

  const { devicePixelRatio: dpr } = window;
  const { clientHeight: h, clientWidth: w } = container;

  const width = nextMultiple(scale, w);
  const height = nextMultiple(scale, h);

  canvas.width = (width * dpr) / scale;
  canvas.height = (height * dpr) / scale;

  canvas.style.transform = `scale(${max(w / canvas.width, h / canvas.height)})`;
}
