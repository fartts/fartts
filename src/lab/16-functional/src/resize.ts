interface ResizeableElement extends HTMLElement {
  width: number;
  height: number;
}

const { max } = Math;
const { devicePixelRatio: dpr } = window;

function nearest(multipleOf = 1, startingAt = 0) {
  return startingAt % multipleOf !== 0
    ? startingAt + (multipleOf - (startingAt % multipleOf))
    : startingAt;
}

export function resize(
  container: HTMLElement,
  target: ResizeableElement,
  scale = 1,
): void {
  const { clientHeight: h, clientWidth: w } = container;

  const width = nearest(scale, w);
  const height = nearest(scale, h);

  target.width = (width * dpr) / scale;
  target.height = (height * dpr) / scale;

  target.style.transform = `scale(${max(w / target.width, h / target.height)})`;
}
