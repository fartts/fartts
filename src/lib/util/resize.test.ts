import resize from './resize';

describe('@fartts/lib/util/resize', () => {
  test.each`
    w      | h      | cw     | ch     | dpr
    ${300} | ${150} | ${391} | ${291} | ${undefined}
    ${300} | ${150} | ${715} | ${475} | ${1}
    ${300} | ${150} | ${800} | ${450} | ${2}
    ${300} | ${150} | ${959} | ${539} | ${3}
  `('resize [$w, $h] to [$cw, $ch] at dpr: $dpr', ({ w, h, cw, ch, dpr }) => {
    const mockCanvas = {
      width: w,
      height: h,
      style: { width: `${w}px`, height: `${h}px` },
    } as HTMLCanvasElement;

    const mockMain = { clientWidth: cw, clientHeight: ch } as HTMLMainElement;

    Object.defineProperty(window, 'devicePixelRatio', {
      writable: true,
      value: dpr,
    });

    window.dispatchEvent(new Event('resize'));

    expect(resize(mockCanvas, mockMain)).toBe(true);
    expect(resize(mockCanvas, mockMain)).toBe(false);

    window.dispatchEvent(new Event('resize'));

    expect(resize(mockCanvas, mockMain)).toBe(true);
    expect(resize(mockCanvas, mockMain)).toBe(false);

    expect(mockCanvas).toMatchSnapshot();
  });
});
