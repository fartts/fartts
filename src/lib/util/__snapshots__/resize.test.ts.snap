// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resize [300, 150] to [391, 291] at dpr: undefined 1`] = `
Object {
  "height": 30,
  "style": Object {
    "height": "300px",
    "width": "400px",
  },
  "width": 40,
}
`;

exports[`resize [300, 150] to [715, 475] at dpr: 1 1`] = `
Object {
  "height": 48,
  "style": Object {
    "height": "480px",
    "width": "720px",
  },
  "width": 72,
}
`;

exports[`resize [300, 150] to [800, 450] at dpr: 2 1`] = `
Object {
  "height": 90,
  "style": Object {
    "height": "450px",
    "width": "800px",
  },
  "width": 160,
}
`;

exports[`resize [300, 150] to [959, 539] at dpr: 3 1`] = `
Object {
  "height": 162,
  "style": Object {
    "height": "540px",
    "width": "960px",
  },
  "width": 288,
}
`;
